src/Test.UnitTests.Sarif.WorkItemFiling/WorkItemFilerTests.cs

```diff
@@ -95,9 +104,9 @@ public async Task WorkItemFiler_FilesWorkItemsOnlyForNewResults()
 
             IEnumerable<ResultGroup> filedWorkItems = await filer.FileWorkItems(LogFilePath);
 
-            // The test file NewAndOldResults.sarif contains 5 results, but only 2 of them
-            // have "baselineState": "new".
-            filedWorkItems.Count().Should().Be(2);
+            // The test file NewAndOldResults.sarif contains 5 results, only 2 of them
+            // have "baselineState": "new", but the "all results" filter will take all of them.
+            filedWorkItems.Count().Should().Be(5);
```

Comment:
>5 [](start = 47, length = 1)

These magic numbers are a poor practice, as you know, they are decoupled from the test artifacts and are difficult to maintain (requiring developers to manually understand/verify whether code changes have properly modified these values. A preferable approach would be to author some test utilities that construct the log file in a parameterized way. Alternately, you could glean your magic values from a separate helper that simply parsed the SarifLog file using alternate, well-established code (in this case, you'd deserialize and just cound the results). These kinds of tests are 'self-healing', i.e., as test inputs change, the validation values update in a reliable way.

Reply:
The problem with writing helpers to construct the test artifacts is that now you have to verify the helpers. I've found problems in such helpers before.

The problem with calculating the expected values is that you end up duplicating the product code logic, and then you're not testing anything. For example, suppose I deserialize the log and write this test code:
```
string logFileContents = ... // Extract text from resource.
var log =JsonConvert.Deserialize<SarifLog>(logFileContents);
int expectedCount = log.Runs[0].Results.Where(r => r.BaselineState == BaselineState.New);
...
filedWorkItems.Should().Be(expectedCount);
```
I just rewrote the product code.

The approach you're suggesting is valuable if there is an _independent_ way of deciding what outcome you expect. If you have your old copy of Code Complete lying around, see the section "Two Algorithms are Better Than One", pp. 33-38, or just skip to the punch line on p. 38, "Use a second algorithm to validate your results". Of course he's not writing about unit tests -- he's really talking about adding the second algorithm to the Debug version of your _product_ code! -- but the idea is the same.

We can talk about this more when you get back.

---
In reply to: [306120394](https://github.com/microsoft/sarif-sdk/pull/1582#discussion_r306120394) [](ancestors = 306120394)

Reply:
You didn't just rewrite the product code. Instead, you utilize a separate area of the product in order to validate the function of fileWorkItems (which flows through an entirely different code path). Assuming that serialization itself isn't broken (which is very unlikely and which should be triggering other tests), you're ok. Now your test is self-healing: if you adjust the test file, a developer doesn't have to go adjust your magic values.

---
In reply to: [306303730](https://github.com/microsoft/sarif-sdk/pull/1582#discussion_r306303730) [](ancestors = 306303730,306120394)

Reply:
No, the reason `FileWorkItems` arrived at the answer it did (2 filed work items, at the time I wrote my previous comment) is that it invoked the `NewResultsFilteringStrategy`, whose code includes precisely the expression `Where(r =>r.BaselineState == BaselineState.New)` that appears in the test code.

Now, having said that -- you could argue that we have separate unit tests for the `NewResultsFilteringStrategy`, so it's ok to rely on its correctness here. That kind of goes against the spirit of a unit test which is to test a method in isolation, independent of its dependencies.

And having said all that, I've now modified the test to use the trivial `AllResultsFilteringStrategy`, and even I am not so principled that I'm worried about its correctness. So I'll make the change in this test. I'm not convinced this is always the right thing to do.

---
In reply to: [306564691](https://github.com/microsoft/sarif-sdk/pull/1582#discussion_r306564691) [](ancestors = 306564691,306303730,306120394)